lnwallet/musig_session_test.go

```diff
@@ -0,0 +1,247 @@
+package lnwallet
+
+import (
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/stretchr/testify/require"
+)
+
+// nodeType is an enum that represents the two nodes in our test harness.
+type nodeType uint8
+
+const (
+	// nodeAlice is the node that initiates the session.
+	nodeAlice nodeType = iota
+
+	// nodeBob is the node that responds to the session.
+	nodeBob
+)
+
+type muSessionHarness struct {
+	aliceCommit *wire.MsgTx
+	bobCommit   *wire.MsgTx
+
+	aliceSession *MusigPairSession
+
+	bobSession *MusigPairSession
+
+	t *testing.T
+}
+
+func (h *muSessionHarness) selectSession(nodeName nodeType) *MusigPairSession {
+	var targetSession *MusigPairSession
+	switch nodeName {
+	case nodeAlice:
+		targetSession = h.aliceSession
+	case nodeBob:
+		targetSession = h.bobSession
+	}
+
+	return targetSession
+}
+
+func (h *muSessionHarness) refreshSession(nodeName nodeType,
+	nextNonce *musig2.Nonces, revoke bool) {
+
+	var session *MusigPairSession
+	switch nodeName {
+	case nodeAlice:
+		session = h.aliceSession
+	case nodeBob:
+		session = h.bobSession
+	}
+
+	var err error
+
+	// If this isn't in response to a revoke, then we just signed, so we'll
+	// refresh our local session with the newly generated verification
+	// nonce.
+	if !revoke {
+		session.LocalSession, err = session.LocalSession.Refresh(
+			nextNonce,
+		)
+	} else {
+		session.RemoteSession, err = session.RemoteSession.Refresh(
+			nextNonce,
+		)
+	}
+	require.NoError(h.t, err)
+}
+
+// SignCommitment signs a new remote commitment. This is equivalent to sending
+// a CommitSig message on the normal LN protocol.
+func (h *muSessionHarness) SignCommitment(nodeName nodeType) *MusigPartialSig {
+	targetSession := h.selectSession(nodeName)
+
+	sig, err := targetSession.RemoteSession.SignCommit(h.bobCommit)
```

Comment:
I think currently, `targetSession` will only ever be Alice's session meaning that signing Bob's `commitment` is correct here. But if `SignCommitment ` is ever called from a future test using bob's nodeName, then it is `h.aliceCommit` that should be the argument to `Sign`

Reply:
same thing goes for a few other places in this file

Reply:
Yep had this flipped, I modified the test to use diff txns for bob and alice, then fixed the ordering here, PTAL.

Also note that the harness here is a bit restricted to just test that things can be ping ponged back n forth. As an example, it combines steps like sig+revoke into one, and has Bob use one method to verify and alice for the other. 

In part 6, we update all the existing link and channel unit tests to work with musig2. This ends up testing a number of other scenarios.